### ISSUE TITLE
twisted.words.services references nonexistent twisted.words.protocols.irc.IRC_NOSUCHCHANNEL

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4915]
|<img alt="devinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| devinj reported|
|-|-|
|Trac ID|trac#4915|
|Type|defect|
|Created|2011-02-25 16:23:56Z|
|Branch|https://github.com/twisted/twisted/tree/irc-service-channelerror-4915|

http://twistedmatrix.com/trac/browser/trunk/twisted/words/service.py#L549

It meant to say ERR_NOSUCHCHANNEL, as seen later in the same method.

Attachments:

* [Twisted.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/134/134c0b3b267df918319e6ec823c789cf712888be/6dcc3d8103a386244fb8a81af01ea21f8f412c8d.diff) (552 bytes) - added by devinj on 2011-02-25 16:24:21Z - Silly tiny diff.
* [Twisted-tests.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/134/134c0b3b267df918319e6ec823c789cf712888be/55085a0bf1e7dc1e371d577337771a7d0368de0c.diff) (2366 bytes) - added by devinj on 2011-02-25 17:37:41Z - Patch to test cases
* [Twisted-better.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/134/134c0b3b267df918319e6ec823c789cf712888be/54bdb281ad2bed1e04bdd63a178baebf4cd49b67.diff) (891 bytes) - added by devinj on 2011-02-27 19:13:13Z - Updated patch, now fixes irc_PRIVMSG as well.
* [Twisted-tests-better.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/134/134c0b3b267df918319e6ec823c789cf712888be/15c8a1bd3e686d28263049ff2103c0d0488f8236.diff) (4021 bytes) - added by devinj on 2011-02-27 19:13:43Z - Updated tests, now with a more twisted style and docstrings.

<details><summary>Searchable metadata</summary>

```
trac-id__4915 4915
type__defect defect
reporter__devinj devinj
priority__normal normal
milestone__ 
branch__branches_irc_service_channelerror_4915 branches/irc-service-channelerror-4915
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1298651036000000 1298651036000000
changetime__1300085200000000 1300085200000000
version__None None
owner__ 
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="devinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|devinj commented|
|-|-|

The attached test cases test whether the IRC server sends the expected response in all the cases where invalid UTF-8 is sent for text data. Other than the case described in this ticket, there's another case that fails, but I don't know what it's supposed to do. That case is irc_PRIVMSG:

If the passed targetName is not valid UTF-8, it apparently wants to send that targetName as part of the response. But the targetName couldn't be decoded. This leads to an UnboundLocalError.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to devinj|
|-|-|

Hi devinj,

Thanks for the patch!  In the future, make sure you add the "review" keyword to the ticket so someone is certain to notice. :)

Some comments about the patch:

  1. All classes and methods should have docstrings.  For test classes and methods, the docstring needs to explain the behavior being verified.
  1. Non-test methods should generally not have an underscore - in Twisted, underscore is reserved for separating a prefix from a dynamic suffix for dynamic dispatch.  So `assert_chokesOnBadBytes` should be `assertChokesOnBadBytes`.  `_mocked_codes` should be `mockedCodes`.
  1. `test_PRIVMSG` is still failing, can you fix it as well?

Thanks again for helping out here.  It's really great to have more eyes on twisted.words :)

**Comment by twisted-trac** [comment]
|<img alt="devinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|devinj commented|
|-|-|

I updated the tests per your suggestions, and I also fixed test_PRIVMSG. Turns out it was really easy, I just didn't give it more than a glance. The new patches contain all the changes in the old patches, plus the requested additional changes.

I'm not sure if more description is warranted in the tests. Let me know if more rationale-type explanation should be added, and I'll do my best!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [30922]) Branching to 'irc-service-channelerror-4915'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [30923]) Apply patches from the ticket

refs [#4915](https://github.com/twisted/twisted/issues/4915)

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_8"></a>|@lvh commented|
|-|-|

Looks good to me. Merging.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_9"></a>|@lvh set status to `closed`|
|-|-|

(In [31007]) twisted.words.services no longer references nonexistent twisted.words.protocols.irc.IRC_NOSUCHCHANNEL.

Author: devinj
Reviewer: lvh, exarkun
Fixes: [#4915](https://github.com/twisted/twisted/issues/4915)

twisted.words.services is no longer fundamentally wrong, and a bunch of test got added in the related code. Yay!
